libindy_vdr/src/pool/cache/strategy.rs

```diff
@@ -0,0 +1,199 @@
+            Some((ts, v)) => {
+                if current_time < *ts {
+                    Some((*ts, v.clone()))
+                } else {
+                    store_lock.remove(key);
+                    None
+                }
+            }
+            None => None,
+        };
+        // update the timestamp if the entry is still valid
+        if let Some((_, ref v)) = get_res {
+            store_lock.re_order(key, current_time + v.expire_offset);
+        }
+        get_res.map(|(_, v)| v.value)
+    }
+    async fn remove(&mut self, key: &K) -> Option<V> {
+        self.store.lock().await.remove(key).map(|(_, v)| v.value)
+    }
+
+    async fn insert(&mut self, key: K, value: V, custom_exp_offset: Option<u128>) -> Option<V> {
+        let mut store_lock = self.store.lock().await;
+        let current_ts = SystemTime::now()
+            .duration_since(self.create_time)
+            .unwrap()
+            .as_millis();
+
+        // remove expired entries
+        while store_lock.len() > 0
+            && store_lock
+                .get_first_key_value()
+                .map(|(_, ts, _)| ts.clone() < current_ts)
+                .unwrap_or(false)
+        {
+            store_lock.remove_first();
+        }
+
+        // remove the oldest item if the cache is still full
+        if store_lock.len() >= self.capacity && store_lock.get(&key).is_none() {
+            // remove the oldest item
+            let removal_key = store_lock.get_first_key_value().map(|(k, _, _)| k.clone());
+            if let Some(removal_key) = removal_key {
+                store_lock.remove(&removal_key);
+            }
+        };
+
+        let exp_offset = custom_exp_offset.unwrap_or(self.expire_after);
+        store_lock
+            .insert(
+                key,
+                TTLCacheItem {
+                    value: value,
+                    expire_offset: exp_offset,
+                },
+                current_ts + exp_offset,
+            )
+            .map(|v| v.value)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+
+    use std::thread;
+
+    use super::*;
+    use crate::pool::cache::{storage::OrderedHashMap, Cache};
+    use futures_executor::block_on;
+
+    #[rstest]
+    fn test_cache_ttl() {
+        let cache = Cache::new(CacheStrategyTTL::new(2, 5, None, None));
+        let cache_location = "test_fs_cache_ttl";
+        let tree = sled::open(cache_location)
+            .unwrap()
+            .open_tree(cache_location)
+            .unwrap();
+        let storage: OrderedHashMap<String, u128, TTLCacheItem<String>> = OrderedHashMap::new(tree);
+        let fs_cache = Cache::new(CacheStrategyTTL::new(2, 5, Some(storage), None));
+        let caches = vec![cache, fs_cache];
+        block_on(async {
+            for cache in caches {
+                cache
+                    .insert("key".to_string(), "value".to_string(), None)
+                    .await;
+                thread::sleep(std::time::Duration::from_millis(1));
```

Comment:
I'm not sure why these sleeps are necessary, is the cache not updated by the time insert completes?

Reply:
They're not really necessary, I just put them there for testing because otherwise all the values in the BtreeMap or sled storage would be keyed under the same millisecond